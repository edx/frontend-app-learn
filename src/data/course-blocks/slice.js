/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

const blocksSlice = createSlice({
  name: 'blocks',
  initialState: {
    fetchState: null,
    root: null,
    blocks: {},
  },
  reducers: {
    fetchCourseBlocksRequest: (draftState) => {
      draftState.fetchState = 'loading';
    },
    fetchCourseBlocksSuccess: (draftState, { payload }) => ({
      ...payload,
      fetchState: 'loaded',
      loaded: true,
    }),
    fetchCourseBlocksFailure: (draftState) => {
      draftState.fetchState = 'failed';
    },

    // Handles sequence metadata updates
    // Handles completion checks
    fetchBlockMetadataRequest: (draftState, action) => {
      const { blockId } = action.payload;
      if (!draftState.blocks[blockId]) {
        draftState.blocks[blockId] = {};
      }
      draftState.blocks[blockId].fetchState = 'loading';
    },
    fetchBlockMetadataSuccess: (draftState, action) => {
      const { blockId, metadata, relatedBlocksMetadata } = action.payload;
      if (!draftState.blocks[blockId]) {
        draftState.blocks[blockId] = {};
      }
      draftState.blocks[blockId] = {
        ...draftState.blocks[blockId],
        ...metadata,
        fetchState: 'loaded',
        loaded: true,
      };
      if (relatedBlocksMetadata) {
        relatedBlocksMetadata.forEach((blockMetadata) => {
          if (draftState.blocks[blockMetadata.id] === undefined) {
            draftState.blocks[blockMetadata.id] = {};
          }
          draftState.blocks[blockMetadata.id] = {
            ...draftState.blocks[blockMetadata.id],
            ...blockMetadata,
          };
        });
      }
    },
    fetchBlockMetadataFailure: (draftState, action) => {
      const { blockId } = action.payload;
      if (!draftState.blocks[blockId]) {
        draftState.blocks[blockId] = {};
      }
      draftState.blocks[blockId].fetchState = 'failure';
    },

    // Handles bookmarking
    // TODO HANDLES UPDATES TO POSITION in an optimistic way
    updateBlockRequest: (draftState, action) => {
      const { blockId, propertyToUpdate, updateValue } = action.payload;
      const updateStateKey = `${propertyToUpdate}UpdateState`;
      if (!draftState.blocks[blockId]) {
        draftState.blocks[blockId] = {};
      }
      draftState.blocks[blockId][updateStateKey] = 'loading';
      draftState.blocks[blockId][propertyToUpdate] = updateValue;
    },
    updateBlockSuccess: (draftState, action) => {
      const { blockId, propertyToUpdate, updateValue } = action.payload;
      const updateStateKey = `${propertyToUpdate}UpdateState`;
      if (!draftState.blocks[blockId]) {
        draftState.blocks[blockId] = {};
      }
      draftState.blocks[blockId][updateStateKey] = 'updated';
      draftState.blocks[blockId][propertyToUpdate] = updateValue;
    },
    updateBlockFailure: (draftState, action) => {
      const { blockId, propertyToUpdate, initialValue } = action.payload;
      const updateStateKey = `${propertyToUpdate}UpdateState`;
      if (!draftState.blocks[blockId]) {
        draftState.blocks[blockId] = {};
      }
      draftState.blocks[blockId][updateStateKey] = 'failed';
      draftState.blocks[blockId][propertyToUpdate] = initialValue;
    },
  },
});

export const {
  fetchCourseBlocksRequest,
  fetchCourseBlocksSuccess,
  fetchCourseBlocksFailure,
  fetchBlockMetadataRequest,
  fetchBlockMetadataSuccess,
  fetchBlockMetadataFailure,
  updateBlockRequest,
  updateBlockSuccess,
  updateBlockFailure,
} = blocksSlice.actions;

export const { reducer } = blocksSlice;
