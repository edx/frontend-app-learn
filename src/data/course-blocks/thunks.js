import { logError } from '@edx/frontend-platform/logging';
import {
  fetchCourseBlocksRequest,
  fetchCourseBlocksSuccess,
  fetchCourseBlocksFailure,
  fetchBlockMetadataRequest,
  fetchBlockMetadataSuccess,
  fetchBlockMetadataFailure,
  updateBlockRequest,
  updateBlockSuccess,
  updateBlockFailure,
} from './slice';
import {
  getCourseBlocks,
  getSequenceMetadata,
  getBlockCompletion,
  updateSequencePosition,
  createBookmark,
  deleteBookmark,
} from './api';

export function fetchCourseBlocks(courseUsageKey) {
  return async (dispatch) => {
    dispatch(fetchCourseBlocksRequest(courseUsageKey));
    try {
      const courseBlocks = await getCourseBlocks(courseUsageKey);
      dispatch(fetchCourseBlocksSuccess(courseBlocks));
    } catch (error) {
      logError(error);
      dispatch(fetchCourseBlocksFailure(courseUsageKey));
    }
  };
}

export function fetchSequenceMetadata(sequenceBlockId) {
  return async (dispatch) => {
    dispatch(fetchBlockMetadataRequest({ blockId: sequenceBlockId }));
    try {
      const sequenceMetadata = await getSequenceMetadata(sequenceBlockId);
      dispatch(fetchBlockMetadataSuccess({
        blockId: sequenceBlockId,
        metadata: sequenceMetadata,
        relatedBlocksMetadata: sequenceMetadata.items,
      }));
    } catch (error) {
      logError(error);
      dispatch(fetchBlockMetadataFailure({ blockId: sequenceBlockId }));
    }
  };
}

export function checkBlockCompletion(courseUsageKey, sequenceId, unitId) {
  return async (dispatch, getState) => {
    const { courseBlocks } = getState();
    if (courseBlocks.blocks[unitId].complete) {
      return; // do nothing. Things don't get uncompleted after they are completed.
    }
    const commonPayload = { blockId: unitId, fetchType: 'completion' };
    dispatch(fetchBlockMetadataRequest(commonPayload));
    try {
      const isComplete = await getBlockCompletion(courseUsageKey, sequenceId, unitId);
      dispatch(fetchBlockMetadataSuccess({
        ...commonPayload,
        metadata: {
          complete: isComplete,
        },
      }));
    } catch (error) {
      logError(error);
      dispatch(fetchBlockMetadataFailure(commonPayload));
    }
  };
}

export function saveSequencePosition(courseUsageKey, sequenceId, position) {
  return async (dispatch, getState) => {
    const { courseBlocks } = getState();
    const actionPayload = {
      blockId: sequenceId,
      propertyToUpdate: 'position',
      updateValue: position,
      initialValue: courseBlocks.blocks[sequenceId].position,
    };
    dispatch(updateBlockRequest(actionPayload));
    try {
      await updateSequencePosition(courseUsageKey, sequenceId, position);
      dispatch(updateBlockSuccess(actionPayload));
    } catch (error) {
      logError(error);
      dispatch(updateBlockFailure(actionPayload));
    }
  };
}

export function addBookmark(unitId) {
  return async (dispatch) => {
    const actionPayload = {
      blockId: unitId,
      propertyToUpdate: 'bookmarked',
      updateValue: true,
      initialValue: false,
    };
    dispatch(updateBlockRequest(actionPayload));
    try {
      await createBookmark(unitId);
      dispatch(updateBlockSuccess(actionPayload));
    } catch (error) {
      logError(error);
      dispatch(updateBlockFailure(actionPayload));
    }
  };
}

export function removeBookmark(unitId) {
  return async (dispatch) => {
    const actionPayload = {
      blockId: unitId,
      propertyToUpdate: 'bookmarked',
      updateValue: false,
      initialValue: true,
    };
    dispatch(updateBlockRequest(actionPayload));
    try {
      await deleteBookmark(unitId);
      dispatch(updateBlockSuccess(actionPayload));
    } catch (error) {
      logError(error);
      dispatch(updateBlockFailure(actionPayload));
    }
  };
}
