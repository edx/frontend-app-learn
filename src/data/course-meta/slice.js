/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';
import PropTypes from 'prop-types';

const courseMetaSlice = createSlice({
  name: 'course-meta',
  initialState: {
    fetchState: null,
  },
  reducers: {
    fetchCourseMetadataRequest: (draftState) => {
      draftState.fetchState = 'loading';
    },
    fetchCourseMetadataSuccess: (draftState, { payload }) => ({
      fetchState: 'loaded',

      /*
       * NOTE: If you change the data saved here,
       * update the courseMetadataShape below!
       */

      // Course identifiers
      name: payload.name,
      number: payload.number,
      org: payload.org,

      // Enrollment dates
      enrollmentStart: payload.enrollmentStart,
      enrollmentEnd: payload.enrollmentEnd,

      // Course dates
      end: payload.end,
      start: payload.start,

      // User access/enrollment status
      enrollmentMode: payload.enrollment.mode,
      isEnrolled: payload.enrollment.isActive,
      userHasAccess: payload.userHasAccess,
      isStaff: payload.userHasStaffAccess,
      verifiedMode: payload.verifiedMode,

      // Misc
      tabs: payload.tabs,
    }),
    fetchCourseMetadataFailure: (draftState) => {
      draftState.fetchState = 'failed';
    },
  },
});

export const {
  fetchCourseMetadataRequest,
  fetchCourseMetadataSuccess,
  fetchCourseMetadataFailure,
} = courseMetaSlice.actions;

export const { reducer } = courseMetaSlice;

export const courseMetadataShape = PropTypes.shape({
  fetchState: PropTypes.string,
  // Course identifiers
  name: PropTypes.string,
  number: PropTypes.string,
  org: PropTypes.string,

  // Enrollment dates
  enrollmentStart: PropTypes.string,
  enrollmentEnd: PropTypes.string,

  // User access/enrollment status
  enrollmentMode: PropTypes.string,
  isEnrolled: PropTypes.bool,
  userHasAccess: PropTypes.bool,
  isStaff: PropTypes.bool,
  verifiedMode: PropTypes.shape({
    price: PropTypes.number.isRequired,
    currency: PropTypes.string.isRequired,
    currencySymbol: PropTypes.string.isRequired,
    sku: PropTypes.string.isRequired,
    upgradeUrl: PropTypes.string.isRequired,
  }),

  // Course dates
  start: PropTypes.string,
  end: PropTypes.string,

  // Misc
  tabs: PropTypes.arrayOf(PropTypes.shape({
    priority: PropTypes.number,
    slug: PropTypes.string,
    title: PropTypes.string,
    type: PropTypes.string,
    url: PropTypes.string,
  })),
});
