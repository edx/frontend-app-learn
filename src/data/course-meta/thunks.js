/* eslint-disable import/prefer-default-export */
import { logError } from '@edx/frontend-platform/logging';
import {
  fetchCourseMetadataRequest,
  fetchCourseMetadataSuccess,
  fetchCourseMetadataFailure,
} from './slice';
import {
  getCourseMetadata,
} from './api';

export function fetchCourseMetadata(courseUsageKey) {
  return async (dispatch) => {
    dispatch(fetchCourseMetadataRequest({ courseUsageKey }));
    try {
      const courseMetadata = await getCourseMetadata(courseUsageKey);
      dispatch(fetchCourseMetadataSuccess(courseMetadata));
    } catch (error) {
      logError(error);
      dispatch(fetchCourseMetadataFailure({ courseUsageKey }));
    }
  };
}
