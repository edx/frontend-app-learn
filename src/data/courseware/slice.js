/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

export const LOADING = 'loading';
export const LOADED = 'loaded';
export const FAILED = 'failed';

const slice = createSlice({
  name: 'courseware',
  initialState: {
    courseStatus: 'loading',
    courseUsageKey: null,
    sequenceStatus: 'loading',
    sequenceId: null,
  },
  reducers: {
    fetchCourseRequest: (state, { payload }) => {
      state.courseUsageKey = payload.courseUsageKey;
      state.courseStatus = LOADING;
    },
    fetchCourseSuccess: (state, { payload }) => {
      state.courseUsageKey = payload.courseUsageKey;
      state.courseStatus = LOADED;
    },
    fetchCourseFailure: (state, { payload }) => {
      state.courseUsageKey = payload.courseUsageKey;
      state.courseStatus = FAILED;
    },
    fetchSequenceRequest: (state, { payload }) => {
      state.sequenceId = payload.sequenceId;
      state.sequenceStatus = LOADING;
    },
    fetchSequenceSuccess: (state, { payload }) => {
      state.sequenceId = payload.sequenceId;
      state.sequenceStatus = LOADED;
    },
    fetchSequenceFailure: (state, { payload }) => {
      state.sequenceId = payload.sequenceId;
      state.sequenceStatus = FAILED;
    },
  },
});

export const {
  fetchCourseRequest,
  fetchCourseSuccess,
  fetchCourseFailure,
  fetchSequenceRequest,
  fetchSequenceSuccess,
  fetchSequenceFailure,
} = slice.actions;

export const {
  reducer,
} = slice;
