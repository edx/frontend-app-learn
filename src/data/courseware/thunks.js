import { logError } from '@edx/frontend-platform/logging';
import {
  getCourseMetadata,
  getCourseBlocks,
  getSequenceMetadata,
  getBlockCompletion,
  updateSequencePosition,
} from './api';
import {
  addModelsMap, updateModel, updateModels,
} from '../model-store';
import {
  fetchCourseRequest,
  fetchCourseSuccess,
  fetchCourseFailure,
  fetchSequenceRequest,
  fetchSequenceSuccess,
  fetchSequenceFailure,
} from './slice';


export function fetchCourse(courseUsageKey) {
  return async (dispatch) => {
    dispatch(fetchCourseRequest({ courseUsageKey }));
    Promise.all([
      getCourseBlocks(courseUsageKey),
      getCourseMetadata(courseUsageKey),
    ]).then(([
      {
        courses, sections, sequences, units,
      },
      course,
    ]) => {
      dispatch(addModelsMap({
        modelType: 'courses',
        modelsMap: courses,
      }));
      dispatch(updateModel({
        modelType: 'courses',
        model: course,
      }));
      dispatch(addModelsMap({
        modelType: 'sections',
        modelsMap: sections,
      }));
      dispatch(addModelsMap({
        modelType: 'sequences',
        modelsMap: sequences,
      }));
      dispatch(addModelsMap({
        modelType: 'units',
        modelsMap: units,
      }));
      dispatch(fetchCourseSuccess({ courseUsageKey }));
    }).catch((error) => {
      logError(error);
      dispatch(fetchCourseFailure({ courseUsageKey }));
    });
  };
}

export function fetchSequence(sequenceId) {
  return async (dispatch) => {
    dispatch(fetchSequenceRequest({ sequenceId }));
    try {
      const { sequence, units } = await getSequenceMetadata(sequenceId);
      dispatch(updateModel({
        modelType: 'sequences',
        model: sequence,
      }));
      dispatch(updateModels({
        modelType: 'units',
        models: units,
      }));
      dispatch(fetchSequenceSuccess({ sequenceId }));
    } catch (error) {
      logError(error);
      dispatch(fetchSequenceFailure({ sequenceId }));
    }
  };
}

export function checkBlockCompletion(courseUsageKey, sequenceId, unitId) {
  return async (dispatch, getState) => {
    const { models } = getState();
    if (models.units[unitId].complete) {
      return; // do nothing. Things don't get uncompleted after they are completed.
    }

    try {
      const isComplete = await getBlockCompletion(courseUsageKey, sequenceId, unitId);
      dispatch(updateModel({
        modelType: 'units',
        model: {
          id: unitId,
          complete: isComplete,
        },
      }));
    } catch (error) {
      logError(error);
    }
  };
}

export function saveSequencePosition(courseUsageKey, sequenceId, position) {
  return async (dispatch, getState) => {
    const { models } = getState();
    const initialPosition = models.sequences[sequenceId].position;
    // Optimistically update the position.
    dispatch(updateModel({
      modelType: 'sequences',
      model: {
        id: sequenceId,
        position,
      },
    }));
    try {
      await updateSequencePosition(courseUsageKey, sequenceId, position);
      // Update again under the assumption that the above call succeeded, since it doesn't return a
      // meaningful response.
      dispatch(updateModel({
        modelType: 'sequences',
        model: {
          id: sequenceId,
          position,
        },
      }));
    } catch (error) {
      logError(error);
      dispatch(updateModel({
        modelType: 'sequences',
        model: {
          id: sequenceId,
          position: initialPosition,
        },
      }));
    }
  };
}
