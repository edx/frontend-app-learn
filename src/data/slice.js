/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit';

export const LOADING = 'loading';
export const LOADED = 'loaded';
export const FAILED = 'failed';

const slice = createSlice({
  name: 'courseware',
  initialState: {
    courseStatus: 'loading',
    courseId: null,
    sequenceStatus: 'loading',
    sequenceId: null,
  },
  reducers: {
    fetchCourseRequest: (state, { payload }) => {
      state.courseId = payload.courseId;
      state.courseStatus = LOADING;
    },
    fetchCourseSuccess: (state, { payload }) => {
      state.courseId = payload.courseId;
      state.courseStatus = LOADED;
    },
    fetchCourseFailure: (state, { payload }) => {
      state.courseId = payload.courseId;
      state.courseStatus = FAILED;
    },
    fetchSequenceRequest: (state, { payload }) => {
      state.sequenceId = payload.sequenceId;
      state.sequenceStatus = LOADING;
    },
    fetchSequenceSuccess: (state, { payload }) => {
      state.sequenceId = payload.sequenceId;
      state.sequenceStatus = LOADED;
    },
    fetchSequenceFailure: (state, { payload }) => {
      state.sequenceId = payload.sequenceId;
      state.sequenceStatus = FAILED;
    },
  },
});

export const {
  fetchCourseRequest,
  fetchCourseSuccess,
  fetchCourseFailure,
  fetchSequenceRequest,
  fetchSequenceSuccess,
  fetchSequenceFailure,
} = slice.actions;

export const {
  reducer,
} = slice;
