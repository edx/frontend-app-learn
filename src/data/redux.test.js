import axios from 'axios'; // eslint-disable-line import/no-extraneous-dependencies
import { configureStore } from '@reduxjs/toolkit';
import { Factory } from 'rosie';
import MockAdapter from 'axios-mock-adapter';

import {
  getAuthenticatedHttpClient,
  getAuthenticatedUser,
} from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';
import { logError } from '@edx/frontend-platform/logging';

import { fetchCourse } from './thunks';
import { reducer as coursewareReducer } from './slice';
import { reducer as modelsReducer } from '../model-store';

import './__factories__/courseMetadata.factory';
import './__factories__/courseBlocks.factory';

jest.mock('@edx/frontend-platform/logging', () => ({ logError: jest.fn() }));
jest.mock('@edx/frontend-platform/auth');

const axiosMock = new MockAdapter(axios);
getAuthenticatedHttpClient.mockReturnValue(axios);
getAuthenticatedUser.mockReturnValue({ username: 'edx' });

// Helper, that is used to forcibly finalize all promises
// in thunk before running matcher against state.
const executeThunk = async (thunk, dispatch) => {
  await thunk(dispatch);
  await new Promise(setImmediate);
};

describe('Test thunks', () => {
  let store;

  beforeEach(() => {
    axiosMock.reset();
    logError.mockReset();

    store = configureStore({
      reducer: {
        models: modelsReducer,
        courseware: coursewareReducer,
      },
    });
  });

  describe('Test fetchCourse', () => {
    const courseBlocksUrlRegExp = new RegExp(`${getConfig().LMS_BASE_URL}/api/courses/v2/blocks/*`);
    const courseBaseUrl = `${getConfig().LMS_BASE_URL}/api/courseware/course`;

    it('Should fail to fetch if error occurs', async () => {
      const courseUrl = `${courseBaseUrl}/courseId`;

      axiosMock.onGet(courseUrl).networkError();
      axiosMock.onGet(courseBlocksUrlRegExp).networkError();

      await executeThunk(fetchCourse('courseId'), store.dispatch);

      const state = store.getState();
      expect(state.courseware.courseStatus).toEqual('failed');
      expect(state).toMatchSnapshot();
    });

    it('Should fetch, normalize, and save metadata, but with denied status', async () => {
      const courseMetadata = Factory.build('courseMetadata', {
        can_load_courseware: {
          has_access: false,
        },
      });
      const courseBlocks = Factory.build('courseBlocks', { courseId: courseMetadata.id });

      const courseUrl = `${courseBaseUrl}/${courseMetadata.id}`;

      axiosMock.onGet(courseUrl).reply(200, courseMetadata);
      axiosMock.onGet(courseBlocksUrlRegExp).reply(200, courseBlocks);

      await executeThunk(fetchCourse(courseMetadata.id), store.dispatch);

      const state = store.getState();
      expect(state.courseware.courseStatus).toEqual('denied');
      expect(state).toMatchSnapshot();
    });

    it('Should fetch, normalize, and save metadata', async () => {
      const courseMetadata = Factory.build('courseMetadata');
      const courseBlocks = Factory.build('courseBlocks', { courseId: courseMetadata.id });

      const courseUrl = `${courseBaseUrl}/${courseMetadata.id}`;

      axiosMock.onGet(courseUrl).reply(200, courseMetadata);
      axiosMock.onGet(courseBlocksUrlRegExp).reply(200, courseBlocks);

      await executeThunk(fetchCourse(courseMetadata.id), store.dispatch);

      const state = store.getState();
      expect(state.courseware.courseStatus).toEqual('loaded');
      expect(state).toMatchSnapshot();
    });
  });
});
