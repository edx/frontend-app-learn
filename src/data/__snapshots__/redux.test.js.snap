// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test thunks Test fetchCourse Should fail to fetch if error occurs 1`] = `
Object {
  "courseware": Object {
    "courseId": "courseId",
    "courseStatus": "failed",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {},
}
`;

exports[`Test thunks Test fetchCourse Should fetch, normalize, and save metadata 1`] = `
Object {
  "courseware": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course_2",
    "courseStatus": "loaded",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courses": Object {
      "course-v1:edX+DemoX+Demo_Course_2": Object {
        "canLoadCourseware": Object {
          "additionalContextUserMessage": null,
          "developerMessage": null,
          "errorCode": null,
          "hasAccess": true,
          "userFragment": null,
          "userMessage": null,
        },
        "canShowUpgradeSock": false,
        "celebrations": undefined,
        "contentTypeGatingEnabled": false,
        "courseExpiredMessage": null,
        "end": null,
        "enrollmentEnd": null,
        "enrollmentMode": null,
        "enrollmentStart": null,
        "id": "course-v1:edX+DemoX+Demo_Course_2",
        "isEnrolled": null,
        "isStaff": false,
        "marketingUrl": null,
        "notes": Object {
          "enabled": false,
          "visible": true,
        },
        "number": "DemoX",
        "offerHtml": null,
        "org": "edX",
        "sectionIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd7",
        ],
        "showCalculator": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "priority": 0,
            "slug": "courseware",
            "title": "Course",
            "type": "courseware",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_2/course/",
          },
          Object {
            "priority": 1,
            "slug": "discussion",
            "title": "Discussion",
            "type": "discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_2/discussion/forum/",
          },
          Object {
            "priority": 2,
            "slug": "wiki",
            "title": "Wiki",
            "type": "wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_2/course_wiki",
          },
          Object {
            "priority": 3,
            "slug": "progress",
            "title": "Progress",
            "type": "progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_2/progress",
          },
          Object {
            "priority": 4,
            "slug": "instructor",
            "title": "Instructor",
            "type": "instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_2/instructor",
          },
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd8",
        "verifiedMode": Object {
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "sections": Object {
      "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd7": Object {
        "courseId": "course-v1:edX+DemoX+Demo_Course_2",
        "id": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd7",
        "sequenceIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd6",
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd7",
      },
    },
    "sequences": Object {
      "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd6": Object {
        "id": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd6",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd6",
        "sectionId": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd7",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd6",
        "unitIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd5",
        ],
      },
    },
    "units": Object {
      "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd5": Object {
        "graded": false,
        "id": "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd5",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd5",
        "sequenceId": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd6",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd5",
      },
    },
  },
}
`;

exports[`Test thunks Test fetchCourse Should fetch, normalize, and save metadata, but with denied status 1`] = `
Object {
  "courseware": Object {
    "courseId": "course-v1:edX+DemoX+Demo_Course_1",
    "courseStatus": "denied",
    "sequenceId": null,
    "sequenceStatus": "loading",
  },
  "models": Object {
    "courses": Object {
      "course-v1:edX+DemoX+Demo_Course_1": Object {
        "canLoadCourseware": Object {
          "hasAccess": false,
        },
        "canShowUpgradeSock": false,
        "celebrations": undefined,
        "contentTypeGatingEnabled": false,
        "courseExpiredMessage": null,
        "end": null,
        "enrollmentEnd": null,
        "enrollmentMode": null,
        "enrollmentStart": null,
        "id": "course-v1:edX+DemoX+Demo_Course_1",
        "isEnrolled": null,
        "isStaff": false,
        "marketingUrl": null,
        "notes": Object {
          "enabled": false,
          "visible": true,
        },
        "number": "DemoX",
        "offerHtml": null,
        "org": "edX",
        "sectionIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        ],
        "showCalculator": false,
        "start": "2013-02-05T05:00:00Z",
        "tabs": Array [
          Object {
            "priority": 0,
            "slug": "courseware",
            "title": "Course",
            "type": "courseware",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course/",
          },
          Object {
            "priority": 1,
            "slug": "discussion",
            "title": "Discussion",
            "type": "discussion",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/discussion/forum/",
          },
          Object {
            "priority": 2,
            "slug": "wiki",
            "title": "Wiki",
            "type": "wiki",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/course_wiki",
          },
          Object {
            "priority": 3,
            "slug": "progress",
            "title": "Progress",
            "type": "progress",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/progress",
          },
          Object {
            "priority": 4,
            "slug": "instructor",
            "title": "Instructor",
            "type": "instructor",
            "url": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course_1/instructor",
          },
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd4",
        "verifiedMode": Object {
          "currency": "USD",
          "currencySymbol": "$",
          "price": 149,
          "sku": "8CF08E5",
          "upgradeUrl": "http://localhost:18130/basket/add/?sku=8CF08E5",
        },
      },
    },
    "sections": Object {
      "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3": Object {
        "courseId": "course-v1:edX+DemoX+Demo_Course_1",
        "id": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        "sequenceIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        ],
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd3",
      },
    },
    "sequences": Object {
      "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2": Object {
        "id": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "sectionId": "block-v1:edX+DemoX+Demo_Course+type@chapter+block@bcdabcdabcdabcdabcdabcdabcdabcd3",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd2",
        "unitIds": Array [
          "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        ],
      },
    },
    "units": Object {
      "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1": Object {
        "graded": false,
        "id": "block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        "lmsWebUrl": "http://localhost:18000/courses/course-v1:edX+DemoX+Demo_Course/jump_to/block-v1:edX+DemoX+Demo_Course+type@vertical+block@bcdabcdabcdabcdabcdabcdabcdabcd1",
        "sequenceId": "block-v1:edX+DemoX+Demo_Course+type@sequential+block@bcdabcdabcdabcdabcdabcdabcdabcd2",
        "title": "bcdabcdabcdabcdabcdabcdabcdabcd1",
      },
    },
  },
}
`;
